File: react-native-poc/ios/RakeenPOC/RakeenCashDrawerModule.swift
import Foundation
import React

/// Standard ESC/POS drawer-kick sequence (pin 2, ~25ms/250ms timing) —
/// same near-universal default already used and documented in
/// ios/App/App/MainViewController.swift's Capacitor bridge. Overridable
/// per `CashDrawerOpenOptions.kickCommandBase64`
/// (react-native-poc/src/platform/cashDrawer.ts) rather than hardcoded
/// with no escape hatch — but not overridden anywhere yet, since no real
/// hardware has ever required different bytes.
private let defaultKickCommand: [UInt8] = [0x1B, 0x70, 0x00, 0x19, 0xFA]

@objc(RakeenCashDrawerModule)
class RakeenCashDrawerModule: NSObject {

    private let transport = NetworkPrinterTransport()
    private let bluetoothTransport = BluetoothPrinterTransport()

    @objc
    static func requiresMainQueueSetup() -> Bool { return false }

    /// options: { target: { transport, host?, port?, bluetoothId? },
    /// kickCommandBase64?: String, operationId: String, timeoutMs: Int }.
    /// `operationId` is accepted but intentionally NOT used for dedup here
    /// -- double-kick protection is enforced JS-side in openCashDrawer()
    /// (react-native-poc/src/platform/cashDrawer.ts), since this native
    /// method has no concept of "the same logical operation" across calls;
    /// it only ever executes exactly one kick per invocation. Feature
    /// Parity Pass -- Bluetooth/USB: most real setups wire the drawer
    /// through the receipt printer's own RJ11 port, so the kick travels
    /// over whatever transport that printer uses -- a Bluetooth printer's
    /// drawer kick goes out over the same BLE write path as its print
    /// jobs. iOS has no USB drawer path (no USB printer path either).
    @objc
    func open(_ options: NSDictionary, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        guard let target = options["target"] as? NSDictionary else {
            resolve(["ok": false, "error": "INVALID_TARGET"])
            return
        }
        let bytes: Data
        if let overrideBase64 = options["kickCommandBase64"] as? String,
           let overrideBytes = Data(base64Encoded: overrideBase64) {
            bytes = overrideBytes
        } else {
            bytes = Data(defaultKickCommand)
        }
        let timeoutMs = (options["timeoutMs"] as? NSNumber)?.intValue ?? 8000
        let transportKind = target["transport"] as? String ?? "network"

        switch transportKind {
        case "bluetooth":
            guard let bluetoothId = target["bluetoothId"] as? String else {
                resolve(["ok": false, "error": "INVALID_TARGET"])
                return
            }
            bluetoothTransport.send(bytes: bytes, peripheralId: bluetoothId, timeoutMs: timeoutMs) { ok, errorDetail in
                if ok {
                    resolve(["ok": true])
                } else {
                    resolve(["ok": false, "error": "PRINTER_CONNECTION_FAILED", "errorDetail": errorDetail ?? "unknown_error"])
                }
            }
        case "usb":
            resolve(["ok": false, "error": "TRANSPORT_NOT_SUPPORTED", "errorDetail": "usb_not_supported_on_ios"])
        default:
            guard let host = target["host"] as? String, let portNumber = target["port"] as? NSNumber else {
                resolve(["ok": false, "error": "INVALID_TARGET"])
                return
            }
            let port = UInt16(truncating: portNumber)
            transport.send(bytes: bytes, host: host, port: port) { ok, errorDetail, diagnostics in
                if ok {
                    resolve(["ok": true, "diagnostics": diagnostics])
                } else {
                    resolve(["ok": false, "error": "PRINTER_CONNECTION_FAILED", "errorDetail": errorDetail ?? "unknown_error", "diagnostics": diagnostics])
                }
            }
        }
    }

    @objc
    func getCapabilities(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        resolve(["supported": true])
    }
}

File: react-native-poc/android/app/src/main/java/com/rakeenpoc/RakeenCashDrawerModule.kt
package com.rakeenpoc

import android.util.Base64
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap

/** Standard ESC/POS drawer-kick sequence (pin 2, ~25ms/250ms timing) --
 *  same near-universal default as ios/RakeenPOC/RakeenCashDrawerModule.swift
 *  and the Capacitor project's MainViewController.swift. Overridable via
 *  kickCommandBase64 (see react-native-poc/src/platform/cashDrawer.ts) but
 *  never overridden anywhere yet -- no real hardware has required
 *  different bytes. */
private val DEFAULT_KICK_COMMAND = byteArrayOf(0x1B, 0x70, 0x00, 0x19.toByte(), 0xFA.toByte())

class RakeenCashDrawerModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    private val transport = NetworkPrinterTransport()
    private val bluetoothTransport = BluetoothClassicPrinterTransport(reactContext)
    private val usbTransport = UsbPrinterTransport(reactContext)

    override fun getName() = "RakeenCashDrawerModule"

    /** options: { target: { transport, host?, port?, bluetoothId?, usbAccessoryId? },
     *  kickCommandBase64?: String, operationId: String, timeoutMs: Int }.
     *  `operationId` is accepted but intentionally NOT used for dedup here
     *  -- double-kick protection is enforced JS-side in openCashDrawer()
     *  (react-native-poc/src/platform/cashDrawer.ts); this native method
     *  always executes exactly one kick per invocation. Feature Parity
     *  Pass -- Bluetooth/USB: the kick travels over whichever transport
     *  the configured printer uses, same principle as iOS's
     *  RakeenCashDrawerModule.swift. */
    @ReactMethod
    fun open(options: ReadableMap, promise: Promise) {
        val target = if (options.hasKey("target")) options.getMap("target") else null
        if (target == null) {
            resolveInvalidTarget(promise)
            return
        }
        val bytes = if (options.hasKey("kickCommandBase64") && !options.isNull("kickCommandBase64")) {
            try {
                Base64.decode(options.getString("kickCommandBase64"), Base64.DEFAULT)
            } catch (e: IllegalArgumentException) {
                DEFAULT_KICK_COMMAND
            }
        } else {
            DEFAULT_KICK_COMMAND
        }
        val timeoutMs = if (options.hasKey("timeoutMs")) options.getInt("timeoutMs") else 8000
        val transportKind = if (target.hasKey("transport")) target.getString("transport") else "network"

        when (transportKind) {
            "bluetooth" -> {
                val bluetoothId = if (target.hasKey("bluetoothId")) target.getString("bluetoothId") else null
                if (bluetoothId == null) { resolveInvalidTarget(promise); return }
                bluetoothTransport.send(bluetoothId, bytes) { ok, errorDetail -> resolveOpenResult(promise, ok, errorDetail) }
            }
            "usb" -> {
                val usbAccessoryId = if (target.hasKey("usbAccessoryId")) target.getString("usbAccessoryId") else null
                if (usbAccessoryId == null) { resolveInvalidTarget(promise); return }
                usbTransport.send(usbAccessoryId, bytes, timeoutMs) { ok, errorDetail -> resolveOpenResult(promise, ok, errorDetail) }
            }
            else -> {
                val host = if (target.hasKey("host")) target.getString("host") else null
                val port = if (target.hasKey("port")) target.getInt("port") else -1
                if (host == null || port <= 0) { resolveInvalidTarget(promise); return }
                transport.send(host, port, bytes) { ok, errorDetail -> resolveOpenResult(promise, ok, errorDetail) }
            }
        }
    }

    private fun resolveInvalidTarget(promise: Promise) {
        val result = Arguments.createMap()
        result.putBoolean("ok", false)
        result.putString("error", "INVALID_TARGET")
        promise.resolve(result)
    }

    private fun resolveOpenResult(promise: Promise, ok: Boolean, errorDetail: String?) {
        val result = Arguments.createMap()
        result.putBoolean("ok", ok)
        if (errorDetail != null) {
            result.putString("error", if (errorDetail == "permission_denied") "PERMISSION_DENIED" else "PRINTER_CONNECTION_FAILED")
            result.putString("errorDetail", errorDetail)
        }
        promise.resolve(result)
    }

    @ReactMethod
    fun getCapabilities(promise: Promise) {
        val result = Arguments.createMap()
        result.putBoolean("supported", true)
        promise.resolve(result)
    }
}
